### ISSUE TITLE
Bazel/Python: AttributeError: 'NoneType' object has no attribute 'decode'

### ISSUE THREAD

**Issue by BenjaminLawson** [google/emboss#110]
The front end compiler is failing on ` pkgutil.get_data(path, "error_examples").decode("utf-8"))` when run with Bazel. I am guessing this is due to the Bazel sandbox not having error_examples, but I'm not sure.

```
benlawson@benlawson:~/pigweed$ bazel test //pw_bluetooth:emboss_test
INFO: Analyzed target //pw_bluetooth:emboss_test (63 packages loaded, 3264 targets configured).
ERROR: /usr/local/google/home/benlawson/pigweed/pw_bluetooth/BUILD.bazel:164:18: Action pw_bluetooth/public/pw_bluetooth/l2cap_frames.emb.ir failed: (Exit 1): emboss_front_end failed: error executing Action command (from target //pw_bluetooth:_emboss_l2cap_frames_ir) 
  (cd /usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed && \
  exec env - \
  bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end pw_bluetooth/public/pw_bluetooth/l2cap_frames.emb '--output-file=bazel-out/k8-fastbuild/bin/pw_bluetooth/public/pw_bluetooth/l2cap_frames.emb.ir' '--import-dir=')
# Configuration: a037c77c201a8c5f95a4c42869d51b86fd83192b95f0c94c91ca81a06d61e1ce
# Execution platform: @@local_config_platform//:host

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Traceback (most recent call last):
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/pigweed/../com_google_emboss/compiler/front_end/emboss_front_end.py", line 173, in <module>
    sys.exit(main(_parse_command_line(sys.argv)))
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/pigweed/../com_google_emboss/compiler/front_end/emboss_front_end.py", line 133, in main
    ir, debug_info, errors = glue.parse_emboss_file(
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/glue.py", line 231, in parse_emboss_file
    ir, debug_info, errors = only_parse_emboss_file(file_name, file_reader)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/glue.py", line 266, in only_parse_emboss_file
    module, module_debug_info, errors = parse_module(file_to_parse,
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/glue.py", line 199, in parse_module
    return parse_module_text(source_code, file_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/glue.py", line 154, in parse_module_text
    parse_result = parser.parse_module(tokens)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/parser.py", line 122, in parse_module
    return _load_module_parser().parse(tokens)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/util/simple_memoizer.py", line 66, in _memoized
    cache[args] = f(*args)
                  ^^^^^^^^
  File "/usr/local/google/home/benlawson/.cache/bazel/_bazel_benlawson/bf871a10d2f2ee7d83e80a1cc326214a/sandbox/linux-sandbox/53/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/parser.py", line 109, in _load_module_parser
    pkgutil.get_data(path, "error_examples").decode("utf-8"))
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'decode'
```

**Comment by BenjaminLawson** [comment]
Interestingly, the Bazel rule worked fine in a different project. Maybe the bazel settings are different.

**Comment by BenjaminLawson** [comment]
`error_examples` is declared as a data dep in the parser target, so maybe it's not a sandbox issue:
```
py_library(
    name = "parser",
    srcs = ["parser.py"],
    data = [
        "error_examples",
    ],
```

**Comment by BenjaminLawson** [comment]
The Python version being used is `Python 3.8.10+chromium.24`.

**Comment by BenjaminLawson** [comment]
The version of Bazel being used is
Build label: 7.0.2
Build target: @@//src/main/java/com/google/devtools/build/lib/bazel:BazelServer
Build time: Thu Jan 25 16:13:35 2024 (1706199215)
Build timestamp: 1706199215
Build timestamp as int: 1706199215

**Comment by BenjaminLawson** [comment]
I noticed that `data` no longer appears to be a documented parameter of py_library: https://bazel.build/reference/be/python#py_library

**Comment by tpudlik** [comment]
`data` is one of the "common" attributes, https://bazel.build/reference/be/common-definitions#typical-attributes.

The documentation doesn't list `srcs` or `deps` either. Not terribly useful :/

Is `error_examples` another target in the BUILD file? Of what type?

**Comment by reventlov** [comment]
`error_examples` is just a file.  `pkgutil.get_data` uses the same mechanism as Python's module loader to actually find the file, so in theory it should work if `import`s are working and Bazel hasn't dropped the file from the sandbox or stuffed it somewhere strange.

(Does Bazel 7 stick `data` dependencies somewhere different from code dependencies now?  I need to set up a Bazel 7 environment and test.)

**Comment by tpudlik** [comment]
Don't you need to use [runfiles](https://github.com/bazelbuild/rules_python/tree/main/python/runfiles) to access files specified via `data`? I've never seen `pkgutil.get_data` used for this.

**Comment by reventlov** [comment]
It's a little bit of an artifact from when I open sourced this (the internal version used a different-but-roughly-equivalent system), but `pkgutil.get_data` is more or less what is necessary to properly grab the file when running as raw Python with the `embossc` driver.

**Comment by BenjaminLawson** [comment]
`data` was a documented parameter in 6.4.0: https://bazel.build/versions/6.4.0/reference/be/python#py_library

**Comment by BenjaminLawson** [comment]
I see the error_examples file in the Bazel sandbox at `sandbox/linux-sandbox/8/execroot/pigweed/bazel-out/k8-opt-exec-ST-13d3ddad9198/bin/external/com_google_emboss/compiler/front_end/emboss_front_end.runfiles/com_google_emboss/compiler/front_end/error_examples`.

**Comment by BenjaminLawson** [comment]
I think it would be ideal to avoid runfiles as we would like this script to work without Bazel too.

**Comment by BenjaminLawson** [comment]
I created a utility for getting data with pkgutil or runfiles, whichever works. It seems like it needs to be used everywhere a data file is read.
```
import pkgutil
import sys

try:
  from rules_python.python.runfiles import runfiles
except ImportError:
  pass

def get_data(path, filename):
  data = None
  # When running with Bazel, this will fail because runfiles needs to be used to
  # read data files.
  data_bytes = pkgutil.get_data(path, filename)
  if data_bytes != None:
    data = data_bytes.decode("utf-8")
  elif "rules_python.python.runfiles" in sys.modules:
     r = runfiles.Create()
     location = "com_google_emboss/" + path.replace('.', '/') + "/" + filename
     with open(r.Rlocation(location), "r", encoding = "utf-8") as f:
       data = f.read()
  else:
    raise Exception("Failed to read error_examples")
  return data

```

**Comment by BenjaminLawson** [comment]
Here is my Pigweed CL for reproduction/reference: https://pigweed-review.googlesource.com/c/pigweed/pigweed/+/192513

**Comment by reventlov** [comment]
It appears that `importlib.resources` works where `pkgutil` does not.  I should have a PR together later today.

**Comment by reventlov** [comment]
This appears to be fixed by #111, please reopen if you continue to see the same error.
